### PR TITLE
Create `optuna.visualization.matplotlib`.

### PR DESCRIPTION
**Pull request by ytknzw**
This PR proposes to implement ``matplotlib`` backend for plotting functions as ``optuna.visualization.matplotlib``.

## Motivation
To provide another visualization option for users, who have only ``plotly`` now.
See issue #1539 and PR #1707 for the original motivation.

## Description of the changes
Create ``optuna.visualization.matplotlib``. All the functions corresponding to the current Plotly backend are to be implemented in ~PR #1707~ the following PRs.

### CODE REVIEW

File: optuna/visualization/matplotlib/__init__.py
Comment (HideakiImamura) [inline optuna/visualization/matplotlib/__init__.py:1]: I'm not sure, but how about not listing functions that are not yet implemented? When users look at the documentation, they may be confused by the lack of documentation or by the unnatural type of functions.

File: optuna/visualization/matplotlib/_utils.py
Comment (HideakiImamura) [inline optuna/visualization/matplotlib/_utils.py:21]: Could you update the docstring to match the `matplotlib`?
```suggestion
    """Returns whether visualization with `matplotlib` is available or not.

    .. note::

        :mod:`~optuna.visualization` module depends on matplotlib version 3.0.0 or higher. If a
        supported version of matplotlib isn't installed in your environment, this function will return
        :obj:`False`. In such a case, please execute ``$ pip install -U matplotlib>=3.0.0`` to install
        matplotlib.

    Returns:
        :obj:`True` if visualization with `matplotlib` is available, :obj:`False` otherwise.
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! I have several comments. Could you take a look?

File: optuna/visualization/matplotlib/__init__.py
Comment (HideakiImamura) [reply optuna/visualization/matplotlib/__init__.py:1]: I would like to discuss this point with @ytknzw and other reviewers. Thank you!

Comment (c-bata) [reply optuna/visualization/matplotlib/__init__.py:1]: > I'm not sure, but how about not listing functions that are not yet implemented?

I agree with his suggestion.

**Comment by c-bata** [comment]
How about retuning matplotlib's Axes object like `pandas.DataFrame.plot()` function. It makes more customizable (related discussion is https://github.com/optuna/optuna/pull/704).

https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.DataFrame.plot.html

**Comment by c-bata** [comment]
> Some approaches for how we could go about designing the public API. 
> 1. Introduce a “backend” parameter to existing visualization functions.
>     - User must consider the argument in each call, implementation becomes somewhat complicated, and the return value becomes a `Union`.
> 2. Define a global/thread local state. 
>     - User only have to configure it once. Other cons remain. We introduce a global state.
> 3. Introduce `optuna.visualization.matplotlib` to define `plot_*`
>     - User switched between backends by changing the top import statement(s). Pro that we don’t have to keep a global state, code is clearly separated. What functions are implemented in what backend becomes very clear. Return value is not a `Union`.
> 
> I would suggest 3., as you might have noticed. And if so, changes required from this PR are not that large and we could incorporate them here.

https://github.com/optuna/optuna/pull/1707#issuecomment-681335837

In terms of public API's design, I'm also +1 to introduce `optuna.visualization.matplotlib` like this pull request.

File: optuna/visualization/matplotlib/__init__.py
Comment (ytknzw) [reply optuna/visualization/matplotlib/__init__.py:1]: I agree with you two. I didn't think well that part, sorry.

**Comment by ytknzw** [comment]
@HideakiImamura @c-bata 
Thank you for your reviews and comments!
I've made some changes to the codes.

**Comment by ytknzw** [comment]
@c-bata 
https://github.com/optuna/optuna/pull/1756#issuecomment-683383976
> How about retuning matplotlib's Axes object like pandas.DataFrame.plot() function.

Thank you for your suggestion. I assume you mean 'return' not 'retune', right?
My current Matplotlib implementation returns Figure object. For compatibility with the Plotly implementation, I'll work on this in PR #1707 as well.

**Comment by c-bata** [comment]
Sorry for my delayed response :pray:

> Thank you for your suggestion. I assume you mean 'return' not 'retune', right?

Exactly! It was just a typo.

> My current Matplotlib implementation returns Figure object. For compatibility with the Plotly implementation, I'll work on this in PR #1707 as well.

IMO, it's better to type returned value to `matplotlib.Axes` in this pull request to avoid interface breaking changes.
What do you think?

**Comment by toshihikoyanase** [comment]
> How about retuning matplotlib's Axes object like `pandas.DataFrame.plot()` function.

Sounds good. When we released the current visualization functions, some users requested to customize the figures. IMO, it is reasonable to provide such a feature for matplotlib backend.

@ytknzw 
Could you change the return type of `plot_xxx` in this PR, too?

**Comment by ytknzw** [comment]
@c-bata @toshihikoyanase 
Thank you for your suggestions. Yes, I agree with you.
Just changed the return value type to `Axes`.

**Review by HideakiImamura — APPROVED** [review]
LGTM!

**Review by c-bata — APPROVED** [review]
LGTM after passed CI 👍

File: optuna/visualization/matplotlib/_utils.py
Comment (toshihikoyanase) [inline optuna/visualization/matplotlib/_utils.py:11]: A nit, but a single backtick corresponds to italic. Please use double backticks for consistency. 
```suggestion
    """Returns whether visualization with ``matplotlib`` is available or not.
```

File: optuna/visualization/matplotlib/_utils.py
Comment (toshihikoyanase) [inline optuna/visualization/matplotlib/_utils.py:15]: `optuna.visualization.plot_xx` does not depend on Matplotlib. Could you update the module path?

```suggestion
        :mod:`~optuna.visualization.matplotlib` module depends on Matplotlib version 3.0.0 or higher. If a
```

File: optuna/visualization/matplotlib/_utils.py
Comment (toshihikoyanase) [inline optuna/visualization/matplotlib/_utils.py:21]: ditto.
```suggestion
        :obj:`True` if visualization with ``matplotlib`` is available, :obj:`False` otherwise.
```

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your update. I added some small comments. I may be able to work on them as a follow-up task.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your swift action. LGTM. I'm waiting for the CI jobs to merge this PR.

**Pull request merged**
